flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/EmbeddedCompletedCheckpointStore.java

```diff
@@ -44,19 +46,36 @@ private static void throwAlreadyShutdownException(JobStatus status) {
 
     private final int maxRetainedCheckpoints;
 
+    @VisibleForTesting
     public EmbeddedCompletedCheckpointStore() {
         this(1);
     }
 
+    @VisibleForTesting
     public EmbeddedCompletedCheckpointStore(int maxRetainedCheckpoints) {
         this(maxRetainedCheckpoints, Collections.emptyList());
     }
 
     public EmbeddedCompletedCheckpointStore(
             int maxRetainedCheckpoints, Collection<CompletedCheckpoint> initialCheckpoints) {
+        this(
+                maxRetainedCheckpoints,
+                initialCheckpoints,
+                SharedStateRegistry.DEFAULT_FACTORY.create(Executors.directExecutor()));
+    }
+
+    public EmbeddedCompletedCheckpointStore(
+            int maxRetainedCheckpoints,
+            Collection<CompletedCheckpoint> initialCheckpoints,
+            SharedStateRegistry sharedStateRegistry) {
+        super(sharedStateRegistry);
         Preconditions.checkArgument(maxRetainedCheckpoints > 0);
         this.maxRetainedCheckpoints = maxRetainedCheckpoints;
         this.checkpoints.addAll(initialCheckpoints);
+
+        for (CompletedCheckpoint completedCheckpoint : this.checkpoints) {
```

Comment:
That looks wrong. If the `sharedStateRegistry` is restored, we will register checkpoints twice. If the `SharedStateRegistry` was passed from the outside we should assume it's properly populated already.